sdks/python/apache_beam/runners/common.py

```diff
@@ -51,117 +51,150 @@ def receive(self, windowed_value):
-      windowing: windowing properties of the output PCollection(s)
-      context: a DoFnContext to use (deprecated)
-      tagged_receivers: a dict of tag name to Receiver objects
-      logger: a logging module (deprecated)
-      step_name: the name of this step
-      logging_context: a LoggingContext object
-      state: handle for accessing DoFn state
-      scoped_metrics_container: Context switcher for metrics container
-    """
-    self.step_name = step_name
-    self.window_fn = windowing.windowfn
-    self.tagged_receivers = tagged_receivers
-    self.scoped_metrics_container = (scoped_metrics_container
-                                     or ScopedMetricsContainer())
+    signature.do_fn = do_fn
+    signature.process_method = _create_do_fn_method(do_fn, 'process')
+    signature.start_bundle_method = _create_do_fn_method(do_fn, 'start_bundle')
+    signature.finish_bundle_method = _create_do_fn_method(
+        do_fn, 'finish_bundle')
 
-    global_window = GlobalWindow()
+    return signature
 
-    # Need to support multiple iterations.
-    side_inputs = list(side_inputs)
 
-    if logging_context:
-      self.logging_context = logging_context
-    else:
-      self.logging_context = get_logging_context(logger, step_name=step_name)
+class DoFnInvoker(object):
+  """An abstraction that can be used to execute DoFn methods.
 
-    # Optimize for the common case.
-    self.main_receivers = as_receiver(tagged_receivers[None])
+  A DoFnInvoker describes a particular way for invoking methods of a DoFn
+  represented by a given DoFnSignature."""
 
-    # TODO(sourabh): Deprecate the use of context
-    if state:
-      assert context is None
-      self.context = DoFnContext(self.step_name, state=state)
-    else:
-      assert context is not None
-      self.context = context
+  def __init__(self, signature):
+    self.signature = signature
 
-    class ArgPlaceholder(object):
-      def __init__(self, placeholder):
-        self.placeholder = placeholder
+  @staticmethod
+  def create_invoker(
+      signature, use_simple_invoker, context, side_inputs, input_args,
+      input_kwargs):
+    if use_simple_invoker:
+      return SimpleInvoker(signature)
+    else:
+      return PerWindowInvoker(
+          signature, context, side_inputs, input_args, input_kwargs)
 
-    # Stash values for use in dofn_process.
-    self.side_inputs = side_inputs
-    self.has_windowed_inputs = not all(
-        si.is_globally_windowed() for si in self.side_inputs)
+  def invoke_process(self, element, process_output_fn):
+    raise NotImplementedError
 
-    self.args = args if args else []
-    self.kwargs = kwargs if kwargs else {}
-    self.dofn = fn
-    self.dofn_process = fn.process
+  def invoke_start_bundle(self, process_output_fn):
+    defaults = self.signature.start_bundle_method.defaults
+    defaults = defaults if defaults else []
+    args = [self.context if d == core.DoFn.ContextParam else d
```

Comment:
Where is the validation that ContextParam is the only valid pluggable value here?